### ISSUE TITLE
Make React-Script-Editor more secure

### ISSUE THREAD

**Issue by mikezimm** [pnp/sp-dev-fx-webparts#2228]
### Sample

https://github.com/pnp/sp-dev-fx-webparts/tree/main/samples/react-script-editor

### Author(s)

@mikaelsvenson, @techmikael.com

## Overall improvements (Maybe a new version?)

1. Do not allow people to just paste code in a box.
2. Only allow referring to a js or html file in a specific collection on the tenant like a CDN
3. Add analytics to save who adds webpart/updates the settings (for tracability)
4. Optionally add analytics on page loads (who, when, what script file(s) were loaded)

## Suggested additions to help better support the custom code

1. Have an optional property in pane where you can designate the developer/power user
2. Have an optional property in the pane where you have to put a link to documentation
3. Make these something the tenant developer could easily make "required" for all instances or the code will not run
4. Make one or both props visible in the UI so whoever is on the page knows who to call with issues

## Other Tenant setup which could be required to help secure it better:

1. Create designated site collection for the CDN
2. Make "Read All" to everyone
3. Create libraries for power users who want to create code
4. Make it so those users can Create and Modify their code (in their library)
5. Make it so those users can NOT delete versions or files (for traceability)

## What would need to be done with the existing webpart?

1. Create prop in code for Tenant CDN collection
2. Create prop in pane for end user to paste in link to their code
3. Verify pasted link is in the Tenant CDN
4. Create prop in pane for developer/power user name
5. Create prop in code to make that optional or required
6. Create prop in pane for link to documentation
7. Create prop in code to make that optional or required
8. Create prop in code for designated storage of analytics - who updated the webpart
9.  Create prop in code for designated storage of analytics - page load usage (who what when)
10. Verification logic before running any code
11. Way to load and execute the file(s) referenced in # 2
12. Way to save analytics (both parts depending on if the property is not empty

Thoughts?

Things I might be able to help with:
Adding property pane props and webpart props.
Adding the "Owner and Documentation" info into the webpart UI
Verify the link provided is in the CDN
Verify other logic matches before loading the react component

Things I don't know how to do:
Not sure if it's necessary: having more than one file added in the property pane
How to load HTML JS files as text and run "dangerously"

### Are you willing to help?

Yes

**Comment by PathToSharePoint** [comment]
@mikezimm I reviewed your suggestions, and I have a Web Part that does some of this. I'll update it and submit a PR.

Things that I feel won't work:
Designated site collection for the CDN: unless it is hardcoded, I don't see how you would force people to pick a specific URL. The burden of this implementation should not be on the sample, I have an example of text field with pattern (react-ppw-html, currently under PRs) that could be combined with the script editor.
Designate the developer/power user: what is the point, if anyone with edit permissions can modify that property?

Some more info on what I have done so far:
https://blog.pathtosharepoint.com/2020/10/13/transitioning-from-classic-content-editor-web-part-to-modern/

**Comment by PathToSharePoint** [comment]
Let me also ping the author @wobba, I believe the readme has the wrong link.

**Comment by PathToSharePoint** [comment]
@mikezimm I have started working on a code update, this is what the URL control could look like:

![image](https://user-images.githubusercontent.com/40323257/153734880-be2c1c00-a300-4156-9e59-1d9904a97bf5.png)

In the above snippet, the HTML5 pattern attribute is used to validate the location of the script.. [PropertyPaneWrap](https://www.npmjs.com/package/property-pane-wrap) is my wrap utility for the Property Pane.

**Comment by mikezimm** [comment]
@PathToSharePoint , thanks for the reply and comments.

In our case, we have some developers in our Engineering group which are not on the SharePoint team.  Therefore it's harder to expect them to learn/install the entire toolchain of SPFx.  So our goal is to provide an alternative to them just putting it on a classic page somewhere 'off the record'.  ie Make it easier to do the right thing.

## Site Collection CDN:
I was thinking of just creating a 'string' in the strings file (because it's easy to find) that could be blank by default.  

Then using your pattern above, we could just do something like this...
```
pattern:  strings.cdnUrl.length > 0 ? strings.cdnUrl: "/sites[^/+][rest of your pattern]",
```


## Designate the developer/power user:  what is the point
We just want to bake in a way for people to get support from the person who wrote the code.  This would also help our SharePoint team (8 people for 25k + office users) find out who to call if something isn't working.

We would also like to enforce that documentation/contact must be somewhere on the page (like a small footer).  But at least have a name and a valid url for docs helps for support.  I'm sure someone who knows JS could bypass anything like that but again we are just trying to make it easier to do the right thing.


## Analytics for tracking page edits and views:
I do list-based analytics in all my webparts because I don't have the background or rights for Azure insites.  If I were to put that into a sample, I'd use the same method for the CDN verification.  Just have a string property for the Url/List Name to save to and if they both have a value, then it tries to save the data.


Anyway, thanks for the feedback :)

**Comment by mikezimm** [comment]
@PathToSharePoint ,
I was just experimenting, trying to load an html file (or any file) using pnp.js.

I must have something wrong because I can't get it to load a file either by path or filename.

Here's my very basic 1.11 web part I created to test loading.  If you try to test it out, just update the file info in the react component props. 

https://github.com/mikezimm/FileLoad

**Comment by wobba** [comment]
I don't think the suggestions really make it more secure. If you want to lock down a place where html files can be pulled from, then that is very specific and have to be coded into the web part to make it "secure". If you limit to SiteAssets folders in SPO, then anyone can more or less put a malicious file in some SiteAssets folder they have access to - like in a Team's backed site.

Support loading html/script from a URL instead of pasting is certainly doable - and a feature which could be added.

That said, remember that using a script editor web part defeats the whole purpose of security all together. The only real secure way is to review any deployed code and have it built and deployed in an unchangeable way, like building a proper SPFx web part.

I do understand your scenario, but I don't think you can easily generalize it into the web part... and making it "more" secure is theoretical only imo.

**Comment by PathToSharePoint** [comment]
> If you want to lock down a place where html files can be pulled from, then that is very specific and have to be coded into the web part to make it "secure".

Right, I believe that's what @mikezimm is talking about with the CDN string. The input pattern in my initial screenshot needs to be adjusted, and is only meant as a guide for the users without providing any security. I would also expect some kind of control on the CDNs themselves (could be more than one), with publishing restricted to a small group.

**Comment by wobba** [comment]
I think the best approach is to copy the code and make a content editor web part out of it. Same functionality, but you point to a file instead of allowing in-place edit. Having the same file loading capability in the script editor makes sense as well and should be easy to accomplish.

Anything outside of that makes no sense to me in a generic web part.

For a content editor version you could hardcode to load from `/sites/scripteditor/SiteAssets`, and say that's the one generic location to load from. Anything else is hard to make generic, as it would be tenant specific and more complicated to set up rules.

**Comment by PathToSharePoint** [comment]
> Anything outside of that makes no sense to me in a generic web part.

@mikezimm I would agree with @wobba that some of your requirements are very specific, and better discussed here than included in the sample.
For the contact information, I see a couple options:
- Directly include that information in the snippet itself. You could have a snippet template with a footer, reused by all code contributors
- Simply use a text field to pass an email address
- Use a people picker or equivalent. The PnP SPFx controls and the Microsoft Graph Toolkit have components for that.

**Comment by mikezimm** [comment]
@wobba , @PathToSharePoint ,
I think you are both right thinking it is a CE web part.  It's been so long I totally forgot the obvious difference in that CE loads a file where as script editor has the code embedded in the webpart.

You are also correct that some of my 'requirements' were very specific.  I was just trying to explain my thought process on ways to have a little more control, not necessarily to have all the details done for me.

I did gulp serve @PathToSharePoint 's [CE webpart](https://github.com/PathToSharePoint/dangerous-content-web-part) and think that might be the best starting point.

I would not mind doing some of the details (like prop pane, checking urls for a specific site, add footer/author info).

But some things are little beyond my abilities at the moment.
I've always used react but never hooks, no time to learn hooks yet since I only do spfx in my spare time at work :(.  So for me, it's hard to understand in my mind how the fetchSnippet() function works, what I can and can't do.

For instance, I was able to add an html heading element above where the CE code is placed.  It seemed to work but upon further review, it seems like the js in the 'htmlFragment' does not get executed on viewing.  Only after I edit/publish the page.  Then when I refresh, it's back to just the html part again (no js running).

@PathToSharePoint , would it be best to build it as a separate div like I did?  or just add the html as a string to htmlFragment?  I personally would rather not add to the fragment since it would be harder to read/understand/debug/test than actual react.

## My attempt to add content around the CE part of the webpart:
![image](https://user-images.githubusercontent.com/49648086/153986297-2ed1a909-168c-4189-ab80-e65346d33c39.png)

## The result
This is what the page looks like when you first load it.
Only after I Edit the page and save it does it load all the JS to do the stock ticker chart.

![image](https://user-images.githubusercontent.com/49648086/153986315-f934e805-dbc0-4232-b6a3-1d9065d80673.png)

## HTML in the loaded file for the screenshot
For reference, the html file I'm loading just contains the sample JS from the other sample:
```html
<!-- TradingView Widget BEGIN -->
<div class="tradingview-widget-container">
  <div id="tradingview_e7aa0"></div>
  <div class="tradingview-widget-copyright"><a href="https://www.tradingview.com/symbols/NASDAQ-AAPL/" rel="noopener" target="_blank"><span class="blue-text">AAPL Chart</span></a> by TradingView</div>
  <script type="text/javascript" src="https://s3.tradingview.com/tv.js"></script>
  <script type="text/javascript">
  new TradingView.widget(
  {
  "width": 980,
  "height": 610,
  "symbol": "NASDAQ:AAPL",
  "interval": "D",
  "timezone": "Etc/UTC",
  "theme": "light",
  "style": "1",
  "locale": "en",
  "toolbar_bg": "#f1f3f6",
  "enable_publishing": false,
  "allow_symbol_change": true,
  "container_id": "tradingview_e7aa0"
}
  );
  </script>
</div>
<!-- TradingView Widget END -->
```

**Comment by mikezimm** [comment]
## @PathToSharePoint , if you would be willing to:

1. Get basic CE function working (upon reading the page, js inside html file runs),
2. Put placeholder/empty divs for a header & footer (preferably in the react part, not the html)
3. Provide any guidance/requirements on using the placeholders you think I might need

## I'd like to take a stab at:
Prop Pane for all the extras I was looking for
Add the place-holder component from the script editor sample telling the user to fill out the properties.
Url Validation for pre-configured CDN

For the sake of the sample, I'd be happy to make it work with a blank CDN Url and provide instructions how to use a CDN if it was desired.  If a url is supplied, then maybe even add better instructions right in the placeholder as well.

Thoughts?

**Comment by PathToSharePoint** [comment]
@mikezimm I think you have it right with the header and footer. The issue is actually with your snippet and in particular the div id:
`<div id="tradingview_e7aa0"></div>`

Not sure what's going on, but I suspect the Web Part is tweaking the id to prevent duplicates (in case the page contains multiple instances of a same Web Part).

**Comment by wobba** [comment]
@PathToSharePoint if you add things to the property pane, ensure they are chunked and lazy loaded only for the edit experience :)

**Comment by mikezimm** [comment]
@PathToSharePoint , 
For the trading view example and div id, I did what any good self-taught hack would do... I just copied the code from the [react-script-editor](https://github.com/pnp/sp-dev-fx-webparts/tree/main/samples/react-script-editor) sample and put it in an HTML file, then referenced the file in the CEWP.   I didn't try on a classic CEWP because I was just copying script from a script editor web part.


I'd really be curious to better understand how it all works and what limitations there are in SPO when copying things from Classic or even non-SharePoint applications.  Because that is what people will try to do.  
See code on google/stack exchange >> Try to copy

Speaking of which, my original "stretch goal" for a more complex example/demo, was to package up this really cool code-pen.
https://codepen.io/suth_a/pen/ExbyvjK

If anyone knows how to package it into a single HTML file that will work in a Modern CE Web part, I'd be forever thankful ha ha.  
It could also double as a stress reliever when my build fails...

[This was my first attempt at that sample](https://github.com/mikezimm/SecureScript/blob/2022-02-13-Copy-in-SFPx-Example/src/webparts/secureScript/components/tests/ParticalSample.html) but I just get a black box :(.

Picture is what it should look like but animated based on mouse movements.
![image](https://user-images.githubusercontent.com/49648086/154166910-717ad2bc-90c6-4b3c-bf38-487707a30fb3.png)

**Comment by PathToSharePoint** [comment]
If the hardcoded id is from @wobba then what can I say ;-) Joking aside, that seems to confirm the issue is with my Web Part. Back to the drawing board! I'll test the other sample too.

**Comment by wobba** [comment]
Maybe we have to update the sample in the readme file.. I haven't tested it in a while to verify it is still working. But in theory, any html/script tags you used in classic should work with copy/paste in the modern script editor web part.

I say in theory, as the technical implementation is different. The classic output the script/html on the aspx side server side as part of the original DOM. This is not possible in modern as everything has to be injected and eval'd.

**Comment by PathToSharePoint** [comment]
@mikezimm I took a closer look at the samples. Let me explain why I don't expect them to work as is. And then offer a workaround (and then show how it does not achieve your expected result...).

A Web Part page is a shared space. To make things work, each Web Part should stay within its own boundaries and not appropriate the global space. It should also be compatible with lazy loading of the page (since SP 2013 I believe). Some patterns to avoid:
- load scripts globally (especially the jQuery $)
- use id attributes or document.querySelector
- run scripts that expect HTML elements but don't wait for them to be loaded

The samples in this thread break those rules, and I believe that's why they don't work with my sample Web Part. What you need are snippets that are embeddable, similar to embed code offered on some websites.

If you want to embed a full page code, I'd say the safest workaround is to wrap it in an iframe. I was able to get both samples to run with the following construct:

`<iframe width='100%' height='500px' srcdoc='[sample goes here]'></iframe>`

The CodePen sample was not a full success though, you might want to see this:
![image](https://user-images.githubusercontent.com/40323257/154417953-33113504-65fe-4354-8cff-ee9e2c686fdc.png)

**Comment by mikezimm** [comment]
@PathToSharePoint ,
So to get the effect of a CEWP,
Do I just use the script editor one but put something like this in the 'paste your code here' box?
<iframe width='100%' height='500px' srcdoc='https://tenant.sharepoint.com/sites/collection/siteassets/sample.html/'></iframe>

Or would it be better to use the CE Webpart and do the same thing?  Just a little confused.

Oh and thanks for the heads up on the code-pen thing.  I never saw the error and couldn't tell they were using a 3rd party library.

## On the topic letting script editors and CEWP take over the world...

I found our engineering group already started using a CEWP to bring their full page app to SPO from SP16.  I finally was able to figure out what they did and find the magic 10 lines of CSS that made it work.

So I used browser dev tools to see what my own web part might look like with the css and I really liked it.

Take those 10 lines of css, turn that into 5-6 typescript files, at least a couple hundred lines of code and 6 hours later...

And below is a short video of the end result.

I have new respect for the old saying;
"Keep your friends close, but enemies closer".  Although they really aren't my enemies.  They just want to be able to do things quickly and efficiently the way they have always done in the past.


https://www.youtube.com/embed/HXN0vG2ehDQ
[
![image](https://user-images.githubusercontent.com/49648086/154765705-0f2afc16-5e63-4c95-bff3-13367c041b25.png)
](https://www.youtube.com/embed/HXN0vG2ehDQ)

**Comment by mikezimm** [comment]
@PathToSharePoint ,
Just a quick question... 
In the examples we talked about so far, there was no data in SharePoint.

With putting the js in iframes, would the code inside the iframe have any issues accessing SharePoint with the user permissions?  Or does that become an issue?

Thanks

**Comment by PathToSharePoint** [comment]
I don't see why it wouldn't work, but I'd recommend to do a validation test. I assume the SharePoint code is from the same online tenant.

I'd also like to make it clear that I suggested the iframe as a dirty workaround for third party code. If it's your SharePoint and your code, you should write it in a way to avoid globals and it'll work without the iframe.

**Comment by mikezimm** [comment]
@PathToSharePoint , couple other questions... sorry for the bombardment:

Question #1:
Given my goal would be to put the html files in one site collection on the tenant, I made the modifications shown in the snippet below.  It seems to work but I'm just wondering if I'm breaking any rules or best practices using this method?  I've always used pnpjs and never direct spHttpClient call.

```javascript
const tenantCDN = `/sites/ourTenantCDN`;
const snippetURL = `${window.location.origin}${tenantCDN}/_api/web/getFileByServerRelativeUrl('${props.url}')/$value`;
```

Question #2:
To make the iframe method a little more easier for the end user (aka have them just press a toggle and it adds the iframe if it's needed, I realized that if I use your example with srcdoc=' , then I can't have any single quotes in the html file or it crashes because of the single quotes.

That would be challenging for the user because sometimes they might need single quotes.

So my thought was that I could just replace single quotes with \' but that didn't work. At least in my testing.
Here are some just a couple of the numerous regex that I tried with no success.
Any thoughts on how this might be addressed?

```html
  const iframePref = "<iframe width='100%' height='500px' srcdoc='";
  const iframeSuff = "\'</iframe>";
  const noContent = "<div>No content loaded.</div>";
  React.useEffect(() => {
    async function fetchSnippet() {
      //original code for local file
      //const snippetURL = props.ctx.pageContext.web.absoluteUrl + `/_api/web/getFileByServerRelativeUrl('${props.url}')/$value`;
      //code that works when file is in another site collection
      const tenantCDN = `/sites/ourTenantCDN`;
      const snippetURL = `${window.location.origin}${tenantCDN}/_api/web/getFileByServerRelativeUrl('${props.url}')/$value`;
      let htmlFragment: string = (props.url) ? 
      await props.ctx.spHttpClient.get(snippetURL, SPHttpClient.configurations.v1)
      .then((response: SPHttpClientResponse) => {return response.text();})
      :noContent;
      if ( htmlFragment !== noContent ) {
        //Add an iframe around it
        // htmlFragment = htmlFragment.split("'").join("\'");
        htmlFragment = htmlFragment.split("'").join("\\'");
        htmlFragment = `${iframePref}${htmlFragment}${iframeSuff}`;
      }
      const node = document.createRange().createContextualFragment(`${htmlFragment}`);
      setAppendedNode(node);
    }
    fetchSnippet();
  },[props.url, props.displayMode]);
```

**Comment by PathToSharePoint** [comment]
Question 1 => there is no place in #SPFx for admin settings (correct, @VesaJuvonen?). We could imagine a future enhancement where the Web Part has parameters. For now it makes sense to me to either hardcode the values in the Web Part or retrieve them from a secure Microsoft List. I am hardcoding my sample.

Question 2 => Personal opinion: the burden of the iframe should be on the developer. Also, sizing the iframe would be difficult with a generic implementation. As for the quotes, try using a double single quote!

**Comment by mikezimm** [comment]
@PathToSharePoint ,
Q1:  I'm not sure what you mean by 'there is no place in #SPFx for admin settings'.
Are you referring to hard coding a 'tenant cdn' into an example?
If you are talking about an example, we could easily just set the default to empty string and let the person packaging the solution add something if they want.  Then only check for CDN if it has a value.

Q2:  I think the issue with the quotes is that in regular js, you can have a mix of single and double.  But when I just create a giant string, it seems to find the first single quote and consider that the end of that chunk of code.  Not sure how to explain it.  But like if I tried to do something like this, js would not like it:

let comment = 'This is a string with an apostrophy' that will cause an error';

I think the same would happen if I used double quotes if the string of code had any double quotes.  Which is likely a guarantee.

Sizing I see as less of an issue but I'm expecting just about anything built using straight html/js probably is not responsive anyway.

**Comment by mikezimm** [comment]
@wobba , @PathToSharePoint ,
I think I have something working and would like any constructive criticism you can share: 

basically a franken-code built from @wobba 's script editor as a base.
Then [I modified the function](https://github.com/mikezimm/SecureScript/blob/2922-02-21-Load-script-on-render/src/webparts/secureScript/loadDangerous.ts) from @PathToSharePoint and am loading it async to get the fragment.
Then using that instead of the code from the panel.

My previous branch only updated it when the edit panel saved, kind of like @wobba 's original logic.
This branch loads it every render which I think is the way it loaded html files in 2016.
The reason I did that was so that  it would always get the latest html file on every load.

https://github.com/mikezimm/SecureScript/tree/2922-02-21-Load-script-on-render

Anyway, it does seem to work.  I probably broke a few rules but look forward to any guidance you might have.

**Comment by mikezimm** [comment]
@wobba ,
The final challenge is packing the script edtior part with a react Banner or Header component.

Is it possible with SPFx to package your execute script part into a react component like show below that could then get pulled in to the main Webpart Class?  Or does that part need to reside in the main webpart class?

![image](https://user-images.githubusercontent.com/49648086/154983941-af35281e-506c-4e58-a0b5-b0912c15c227.png)

This is the end structure I'm hoping to achieve:
![image](https://user-images.githubusercontent.com/49648086/154984013-58deaeef-6e68-40ec-b2fa-830e51c6c3e3.png)

**Comment by PathToSharePoint** [comment]
@mikezimm no comment on your Web Part as I see it's mainly based on @wobba's code and he'll understand it better.

As planned I have published an update to my "Dangerous Content Web Part" to restrict the files location. I am calling it the "Cherry-Picked Content Web Part" and a first draft is available here:
https://github.com/PathToSharePoint/React-Cherry-Picked-Content

**Comment by wobba** [comment]
@mikezimm You could use SPFx libraries which expose web components, and logic to load them in the script editor webpart so you can use them in the html markup (similar to how we use web components in the Modern Search web parts).

**Comment by PathToSharePoint** [comment]
I have published v0.2.0 of the "Cherry-Picked Content Web Part". It includes a dozen code samples:
https://github.com/PathToSharePoint/React-Cherry-Picked-Content/tree/main/samples

**Comment by wobba** [comment]
Should we close this issue?

**Comment by mikezimm** [comment]
@PathToSharePoint ,
Technically you are correct that the Docs and Contact are tied to the code and not to the web part instance.
So from an example and to keep it simple, I think they could be taken out.

I was considering how to do something like that on my version of the web part (totally re-constructed from pieces of the script editor and your CE web part).

In our version, if we go with that approach, we may just require something hard-coded in the html file that the web-part can easily find and extract.  And if it can't find it, just not load the code or something.

Oh and in my web part, I did get some help and we are actually putting the script editor portion in a new node on the main web part class.  If you want to see how we did it, please let me know and we can get in touch.

@wobba , If you can keep this open at least for a couple days, I'll post a video on what my web currently looks like.  It would be good to see just what can be done if you put your mind to it.

Thanks
Mike

**Comment by wobba** [comment]
Anyone/or myself are free to add a PR which loads up the html/script from a file instead of reading from the web part properties. I  think this is a good extension and should be fairly easy to add.

Anything related to blocking where to read from etc. is just a form of security by obscurity imo and does not make sense. The logic would be very tenant specific and has to rely on proper governance to ensure it works as intended.

**Comment by mikezimm** [comment]
@wobba , @PathToSharePoint ,
I finally got my videos uploaded to you tube in case anyone is interested in seeing what could be done.  Goes beyond a simple sample but can show the power of SPFx and could be useful to someone looking at a more controlled option.

Following up on the previous comments, I agree with @PathToSharePoint  on security by obscurity.  Obscurity does assist.  But you can (to some extent), lock down the code and also raise visibility to the team who supports the entire tenant.

Will that totally prevent someone who is really determined to break the law?  No, but it's like airline security... you probably wouldn't be to popular if you removed all security because you can't stop everything.  You do what is reasonable taking into account all the factors.... performance impacts, inconvenience to end users, security gains, gains in support, technology etc...

This web part does give us some control and a LOT of visibility of what is running on the tenant.

Anyway... here are the links.

This is an introduction - more the big picture of web part and also instructions on how set up the CDN to be more secure.
https://www.youtube.com/watch?v=CMPTP-gBZ5w

This is a walk-through of what I was able to do with the web part...
Including UI options, warnings, blocks, peeling the 'onion' of what code is on the page.
https://www.youtube.com/watch?v=RgLO-CG2mx0

Again thanks for the samples posted here.  It was the inspiration and catalyst for what I did and appreciate the discussions and samples :)

**Comment by wobba** [comment]
> > Anything related to blocking where to read from etc. is just a form of security by obscurity imo
> 
> Allow me to strongly disagree with this statement. Applying permissions to a locatiion is something we do all the time in Teams or SharePoint, and a fundamental security feature.
> 
> > The logic would be very tenant specific and has to rely on proper governance to ensure it works as intended.
> 
> And this statement I strongly agree with 😊`

Let me clarify. I do agree with you, but hard coding a specific path for this purpose and then document that it should be controlled is not a general setting. Does it make sense per tenant and then being implemented based on that tenants governance..sure. I just don't want to make an assumption that /sites/securethis/siteassets is a prereq. And I'm ok with disagreeing as well 🙂

**Comment by mikezimm** [comment]
@wobba , agreed on what you are saying.  My implementation is probably considered more of a 'Solution' which encompasses more than just a web part. 

I do have a code related question about how you were loading and executing the js files....

I thought I saw somewhere that in your example, the scripts are loaded async (in order they are in the code).
Unfortunately I can't find any comments in the code so I'm wondering if I'm dreaming it up or saw it somewhere else.

The reason I am asking my web part is loading 'slow' compared to the proof of concept app we have running in SPO on a classic page with CE webpart.

If that is the case, **was it done for any particular purpose?  Or just the only way it would work?**
Not sure if it's important or not, but in my example, I copied all the external files to our SPO site so it's not on a public CDN.  That might also impact performance a bit.  But I'm more wondering about the async nature of the load.

Here's a link to the file where I have the code from your script editor.  I just moved it into a separate file in my web part.
https://github.com/mikezimm/SecureScript7/blob/main/src/webparts/secureScript7/components/Security20/EvalScripts.ts

# This is how the code loads in the classic CE Webpart:
As you can see, all of the files highlighted in yellow load at the same time.  When you visit the page, there is blinking or anything when it redraws elements/updates css.  It's like native app.

![image](https://user-images.githubusercontent.com/49648086/158699466-3791ee9f-a409-4e42-a4ee-db1b767e5e0c.png)


# This is how it loads using your script execute function:
You can see it loads async and so loading 8-10 scripts will take a lot longer.
In this example, the code loads full screen but we see the sharepoint navigation etc for a half second before it hides it so it is noticeably slower.

![image](https://user-images.githubusercontent.com/49648086/158699663-c7931a3a-a93b-4333-9400-555ec80c83c4.png)

**Comment by mikezimm** [comment]
@wobba , on the performance side, I also noticed that with your script execution, it seems to always be pulling it from the source every time and in our original code, it looks like it is getting it from 'disk cache'.

**Comment by mikezimm** [comment]
@PathToSharePoint ,
Have you noticed with your property pane file picker in the CherryPick webpart that if the library has a 'Link' file ( like shortcut.Url ), the it errors out and it can not fetch the list of files from the library?

**Comment by PathToSharePoint** [comment]
@mikezimm Thanks for the feedback, I'll take a look. I was thinking about filtering the query, to only return files with extension html, htm or text. Note that the idea of the cherry-picked content is that the target libraries should only contain such files, no links or images for example.

**Comment by mikezimm** [comment]
@PathToSharePoint , 
I think filtering on html is a good idea.  It will clean up the dropdown when you have all your other files for the app in the same folder.

I did play around with it in my code trying to find a work around but I am stumped.

I thought the select="Name" at the end of the query was giving an issue because maybe the link file doesn't have a Name but removing that part didn't seem to resolve the issue.

I will be curious to see what you find out.

On the subject of Link files, I have noticed that they mysteriously do not show up in the Teams Files tab.  Maybe they had the same issue there so it may be a clue.

**Comment by PathToSharePoint** [comment]
@mikezimm Let's take the troubleshooting of my solution out of this thread. I'll publish an update, I modified the code:
`/_api/web/lists/getbytitle('" + filesLocation.library + "')/Items?$select=FileLeafRef`
And then the filtering:
`files.map(file => file.FileLeafRef).filter(rf => rf.match(/\.(htm?l|txt)$/i)).sort().map(name => { return { key: name, text: name }; });`
Thanks again!

**Comment by wobba** [comment]
> @wobba , on the performance side, I also noticed that with your script execution, it seems to always be pulling it from the source every time and in our original code, it looks like it is getting it from 'disk cache'.

I'm adding a random param to the URL to force re-load, as that is needed for some code due to smart navigation. If not the code would not re-execute on sub-sequent pages. It could be a toggle of course, but more options makes it harder to use.

**Comment by wobba** [comment]
@PathToSharePoint I always test on a real page using manifest parameter.. just as easy as the workbench imo.

**Comment by mikezimm** [comment]
> > @wobba , on the performance side, I also noticed that with your script execution, it seems to always be pulling it from the source every time and in our original code, it looks like it is getting it from 'disk cache'.
> 
> I'm adding a random param to the URL to force re-load, as that is needed for some code due to smart navigation in SharePoint. It was added to fix that particular reason in v1.0.0.11. If not the code would not re-execute on sub-sequent pages. It could be a toggle of course, but more options makes it harder to use.

@wobba ,
Could you explain the scenarios that cause the reload issue a little more?  Like if it is under certain conditions?

I did a quick google on Smart Navigation and didn't really see anything in the docs for SharePoint Online.  Just a few posts that were pretty old.

For instance, does it come into play when you navigate to that page after clicking on it in the Global Navigation?

Also, in your testing, were you having an issue with the loading or executing of the code?  The reason being I might try to re-factor the fetch part of the code by wrapping all the fetches in a Promise.All so that part doesn't need to be async.  Just the execution from what I recall.  That would probably cut the fetch time by 75% if you have a bunch of references like in one of my examples.

Another possible thought (thinking totally out of the box since I'm not a browser expert)... Are cached files stored in consistent places and would it be possible to always try to load that file from the local hard drive first?  Or is that not possible in SPFx?  I'm guessing not in modern browsers due to security, but was just asking.

Thanks!

Thoughts

**Comment by mikezimm** [comment]
@wobba ,
I'm a little confused by this comment:  " if you have two pages, linking the same script with some different params".  By script are you referring to the script in the script editor or something else?

Also, based on your recollection and knowledge, do you think "it would be obvious" through minimal testing if the toggle can be set to allow caching?  Like 'It either works or doesn't'... or are potentially so many variables it would be more a 'use at your own risk... warranty is voided' option?

**Comment by mikezimm** [comment]
@wobba ,
I had another idea on the caching.  If the issue can be solved by knowing if the current page was previously loaded in the background ( again I don't fully understand what triggers the issue yet )... but in theory could you do something like this:

Create a property on web part (unique id) that is generated when the web part is first added to the page.  Save it to the properties during the on-init only if it currently is null (ie so it will only get generated/saved when you first add the web part to the page).

Then during init, create some variable in the window object that says 'web part Id xyz' was loaded once because the Id is in the window object.

Then somehow use that to determine if all the files should be reloaded?  Such as.... if this web part was previously loaded, then either Load all scripts again or use caching (depending on what is needed).

Anyway, just some ideas.

**Comment by mikezimm** [comment]
@wobba ,
Slightly off topic but maybe related, just trying to understand...

> When you navigate from one SP page to another, then smart navigation happens in the SharePoint React router. There is no full page load, but loading delta wise and updating the DOM.
> 

Is this what happens when you are on a list view and switch views using the view selector?
I noticed that when you do that, the page url does not actually change. It just adds a view id to the parameter.  It's actually kind of annoying if you want to get a clean url to give to someone.

Thanks again for your insights.

**Comment by wobba** [comment]
@mikezimm correct.. or if you navigate between two pages on the same iste. Then the URL changes, but it's typically not a full page re-load.

**Comment by hugoabernier** [comment]
@mikezimm I'll go ahead and close this issue, as it seems to be handled by @PathToSharePoint 's awesome new web part.
